cltk/phonology/orthophonology.py

```diff
@@ -0,0 +1,477 @@
+	    return self.action(phonemes[0])
+
+class WordFinalPhonologicalRule(BasePhonologicalRule):
+	'''
+	A rule applying to the last phoneme in the word.
+	The condition only takes two arguments: before and target.
+	'''
+	def check_environment(self, phonemes, pos):
+	    last = len(phonemes) - 1
+	    return self.condition(phonemes[last - 1], phonemes[last]) if pos == last and len(phonemes) > 1 else False
+
+	def perform_action(self, phonemes, _):
+	    return self.action(phonemes[len(phonemes) - 1])
+
+class InnerPhonologicalRule(BasePhonologicalRule):
+	'''
+	A rule applying to a position inside a word.
+	The before and after arguments to the condition will always be actual phonemes (not None).
+	'''
+	def check_environment(self, phonemes, pos):
+		if pos == 0 or pos == len(phonemes) - 1:
+			return False
+		return self.condition(phonemes[pos - 1], phonemes[pos], phonemes[pos + 1])
+
+class SyllableInitialPhonologicalRule(BasePhonologicalRule):
+	'''
+	A rule applying to the first phoneme in a syllable.
+	Syllable breaks are determined by a simple, language-independent SSP heuristic.
+	The condition only takes two arguments: target and after.
+	'''
+	def check_environment(self, phonemes, pos):
+		if pos == 0 and len(phonemes) > 0:
+			return self.condition(phonemes[pos], phonemes[1]) 
+		elif pos == len(phonemes) - 1:
+			return False
+		# apply simple SSP heuristic to determine if we're at a syllable start
+		elif phonemes[pos - 1].is_more_sonorous(phonemes[pos]) and not phonemes[pos].is_more_sonorous(phonemes[pos + 1]):
+			return self.condition(phonemes[pos], phonemes[pos + 1]) 
+		else:
+			return False
+
+def _check_features(phoneme, feature_values):
+	return reduce(lambda a, b: a or b, [phoneme[type(f)] == f for f in feature_values])
+
+def SimplePhonologicalRule(target, replacement, before=None, after=None):
+	'''
+	Systactic sugar for simple item-specific rules.
+	* target and replacement are phonemes.
+	* before and after, optionally specified, are disjunctive lists of *phonological features*.
+	That is, the rule will fire if ANY of the features listed in e.g. before is present in the preceding phoneme.
+	If neither before nor after are specified, then the rule is UNconditional.  Useful for elsewhere conditions.
+	''' 
+	if before is not None and after is None:
+		cond = lambda b, t, _: t == target and b is not None and _check_features(b, before) 
+	if before is not None and after is not None:
+		cond = lambda b, t, a: t == target and b is not None and _check_features(b, before) and \
+		a is not None and _check_features(a, after)
+	if before is None and after is not None:
+		cond = lambda _, t, a: t == target and a is not None and _check_features(a, after) 
+	if before is None and after is None:
+		cond = lambda _, t, __: t == target
+
+	return PhonologicalRule(cond, lambda _ : replacement)
+
+
+# ------------------- The orthophonology of a language -------------------#
+
+class Orthophonology:

```

Comment:
Is this your own neologism?

Reply:
So, not entirely, but I think when used it's hyphenated.  I've encountered the term in the context of children learning to read.

Reply:
(It may be a bit of an Italianism too.  In Italy when kids have trouble with reading, you march them off to the local _ortofonologo_.)